addons/l10n_eg_edi_eta/models/eta_thumb_drive.py

```diff
@@ -0,0 +1,170 @@
+    _description = 'Thumb drive used to sign invoices in Egypt'
+
+    user_id = fields.Many2one('res.users', required=True, default=lambda self: self.env.user)
+    company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
+    certificate = fields.Binary('ETA Certificate')
+    pin = fields.Char('ETA USB Pin', required=True)
+    access_token = fields.Char(required=True)
+
+    _sql_constraints = [
+        ('user_drive_uniq', 'unique (user_id, company_id)', 'You can only have one thumb drive per user per company!'),
+    ]
+
+    def action_sign_invoice(self, invoice_ids):
+        self.ensure_one()
+        sign_host = self._get_host()
+
+        to_sign_dict = dict()
+        for invoice_id in invoice_ids:
+            eta_invoice = json.loads(invoice_id.l10n_eg_eta_json_doc_id.raw)['request']
+            signed_attrs = self._generate_signed_attrs(eta_invoice, invoice_id.l10n_eg_signing_time)
+            to_sign_dict[invoice_id.id] = base64.b64encode(signed_attrs.dump()).decode()
+
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'action_post_sign_invoice',
+            'params': {
+                'sign_host': sign_host,
+                'access_token': self.access_token,
+                'pin': self.pin,
+                'drive_id': self.id,
+                'invoices': json.dumps(to_sign_dict)
+            }
+        }
+
+    def action_set_certificate_from_usb(self):
+        self.ensure_one()
+        sign_host = self._get_host()
+
+        return {
+            'type': 'ir.actions.client',
+            'tag': 'action_get_drive_certificate',
+            'params': {
+                'sign_host': sign_host,
+                'access_token': self.access_token,
+                'pin': self.pin,
+                'drive_id': self.id
+            }
+        }
+
+    def set_certificate(self, certificate):
+        self.ensure_one()
+        self.certificate = certificate.encode()
+        return True
+
+    def set_signature_data(self, invoices):
+        invoices = json.loads(invoices)
+        for key, value in invoices.items():
+            invoice_id = self.env['account.move'].browse(int(key))
+            eta_invoice_json = json.loads(invoice_id.l10n_eg_eta_json_doc_id.raw)
+
+            cades_bes = self._generate_cades_bes_signature(eta_invoice_json['request'], invoice_id.l10n_eg_signing_time,
+                                                           base64.b64decode(value))
+            signature = base64.b64encode(cades_bes.dump()).decode()
+
+            eta_invoice_json['request']['signatures'] = [{'signatureType': 'I', 'value': signature}]
+            invoice_id.l10n_eg_eta_json_doc_id.raw = json.dumps(eta_invoice_json)
+            invoice_id.l10n_eg_is_signed = True
+        return True
+
+    def _get_host(self):
+        sign_host = self.env['ir.config_parameter'].sudo().get_param('l10n_eg_eta.sign.host', 'http://localhost:8069')
```

Comment:
Hello, just a reminder as we had to solve that for the IOT Box, how are you going to call localhost in http from a page in https? or going to get a certificate for localhost and call it in https?

Pretty sure that the best solution should have been to use an IOT Box and adapt the pricing if needed.

Reply:
I believe this was tested in both FF and Chrome, and `localhost` is subject to an exception in the mixed-content, so it is actually a permitted call. It may not work in Safari (didn't test), but our main target is "typical Windows accountant user".

https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content#loading_locally_delivered_mixed-resources


This means that the `l10n_eg_eta.sign.host` parameter can only contain a localhost-based address, or a fully-validated HTTPS host.

It's different for the IOT-box as it's not running on the loopback address and cannot benefit from this exception.

Reply:
note: the above could be a useful comment to add in the code, as a reminder of this limitation.